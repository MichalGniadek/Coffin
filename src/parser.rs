use crate::{
    ast::{BinOpKind, Expr, Id, Visitor},
    lexer::Token,
    pretty_print::PrettyPrint,
};
use logos::{Lexer, Span, SpannedIter};
use std::iter::Peekable;

pub fn parse<'a>(lexer: Lexer<'a, Token>) {
    let mut parser = Parser {
        lexer: lexer.spanned().peekable(),
        id: Id(0),
        spans: vec![],
    };

    let expr = parser.parse_expr(0);

    println!(
        "{}",
        PrettyPrint::new(parser.spans.clone()).visit_expr(&expr),
    );
}

struct Parser<'a> {
    lexer: Peekable<SpannedIter<'a, Token>>,
    id: Id,
    spans: Vec<Span>,
}

impl<'a> Parser<'a> {
    #![allow(dead_code)]

    fn span(&mut self) -> Span{
        self.lexer.peek().map_or(0..0, |(_, span)| span.clone())
    }

    fn consume(&mut self) -> Id {
        let span = self.span();
        self.spans.push(span);
        self.lexer.next();

        let id = self.id;
        self.id = Id(self.id.0 + 1);
        id
    }

    fn err_consume(&mut self) -> Id {
        const SYNC_TOKENS: [Token; 1] = [Token::RightBrace];

        let mut fin_span = self.span();

        while let Some((token, span)) = self.lexer.peek() {
            if !SYNC_TOKENS.contains(token) {
                fin_span = fin_span.start..span.end;
                self.lexer.next();
            } else {
                break;
            }
        }

        self.spans.push(fin_span.clone());

        let id = self.id;
        self.id = Id(self.id.0 + 1);
        id
    }

    fn eat(&mut self) -> Span {
        let span = self.span();
        self.lexer.next();
        span
    }

    fn peek(&mut self) -> Option<Token> {
        self.lexer.peek().map(|(t, _)| *t)
    }

    fn parse_expr(&mut self, min_binding_power: u8) -> Expr {
        let tree = self.parse_prefix();
        let tree = self.parse_infix(tree, min_binding_power);
        tree
    }

    fn parse_prefix(&mut self) -> Expr {
        match self.peek() {
            Some(Token::Int(i)) => Expr::Int(self.consume(), i),
            Some(Token::LeftBrace) => self.parse_block(),
            Some(Token::RightParen) | Some(Token::Comma) | None => Expr::Error(self.err_consume()),
            Some(_) => Expr::Error(self.err_consume()),
        }
    }

    fn parse_block(&mut self) -> Expr {
        let id = self.consume();
        let mut exprs = Vec::new();

        while let Some(token) = self.peek() {
            // Block ended
            if token == Token::RightBrace {
                break;
            }

            let expr = self.parse_expr(0);

            if let Expr::Error(_) = expr {
                if let Some(Token::RightBrace) = self.peek() {
                    // Expr is an error but it synced to the right brace
                    // so add it to the list and return a Block
                    exprs.push(expr);
                    break;
                } else {
                    // Expr is an error and we can't sync here so return an error
                    return Expr::Error(id);
                }
            } else {
                // Expr is not an error so just add it to the list
                exprs.push(expr);
            }
        }

        // Set the span so that first number represents { and the second represents }
        self.spans[id.0 as usize] = (self.spans[id.0 as usize].start)..(self.eat().end - 1);
        Expr::Block(id, exprs)
    }

    fn parse_infix(&mut self, tree: Expr, min_binding_power: u8) -> Expr {
        // Associativity
        const LEFT: u8 = 0;
        const RIGHT: u8 = 1;

        let mut tree = tree;

        while let Some(token) = self.peek() {
            tree = match token {
                // Token::T if binding_power + associativity > min_binding_power
                Token::Plus if 10 + LEFT > min_binding_power => Expr::Binary(
                    self.consume(),
                    BinOpKind::Add,
                    Box::new(tree),
                    Box::new(self.parse_expr(10)),
                ),
                _ => break,
            }
        }

        tree
    }
}
