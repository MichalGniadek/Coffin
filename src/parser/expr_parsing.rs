use super::Parser;
use crate::{
    ast::{BinOpKind, Expr, Name},
    ast_span,
    error::ParserErrorKind,
    lexer::Token,
};

pub enum ExprResult {
    Correct(Expr),
    PanicMode(Expr),
}
use ExprResult::{Correct, PanicMode};

impl ExprResult {
    pub fn expr(self) -> Expr {
        match self {
            Correct(e) => e,
            PanicMode(e) => e,
        }
    }
    pub fn destruct(self) -> (Expr, bool) {
        match self {
            Correct(e) => (e, false),
            PanicMode(e) => (e, true),
        }
    }
    pub fn map<F: FnOnce(Expr) -> Expr>(self, f: F) -> ExprResult {
        match self {
            Correct(e) => Correct(f(e)),
            PanicMode(e) => PanicMode(f(e)),
        }
    }
}

impl Parser<'_> {
    pub fn parse_expr(&mut self, min_binding_power: Option<u8>) -> ExprResult {
        match self.parse_prefix() {
            Correct(e) => self.parse_infix(e, min_binding_power.unwrap_or(0)),
            PanicMode(e) => PanicMode(e),
        }
    }

    fn parse_prefix(&mut self) -> ExprResult {
        match self.curr_token {
            Token::Int(i) => Correct(Expr::Int(self.consume(), i)),
            Token::Identifier(s) => Correct(Expr::Identifier(Name(self.consume(), s))),
            Token::Let => self.parse_let(),
            Token::LeftBrace => self.parse_block(),
            _ => PanicMode(self.err_consume(
                None,
                ParserErrorKind::ExpectedPrefixToken,
                None,
                &Self::EXPR_SYNC,
            )),
        }
    }

    fn parse_let(&mut self) -> ExprResult {
        let let_id = self.consume_expect(Token::Let);
        
        let mut_id = match self.curr_token {
            Token::Mut => Some(self.consume_expect(Token::Mut)),
            _ => None,
        };

        let name = match self.curr_token {
            Token::Identifier(s) => Name(self.consume(), s),
            _ => {
                return self.err_consume(
                    let_id,
                    ParserErrorKind::expected_identifier(),
                    None,
                    &Self::EXPR_SYNC,
                )
            }
        };

        let eq_id = match self.curr_token {
            Token::Equal => self.consume_expect(Token::Equal),
            _ => {
                return self.err_consume(
                    let_id,
                    ParserErrorKind::ExpectedToken(Token::Equal),
                    None,
                    &Self::EXPR_SYNC,
                )
            }
        };

        self.parse_expr(None).map(|expr| Expr::Let {
            let_id,
            mut_id,
            name,
            eq_id,
            expr: Box::new(expr),
        })
    }

    fn parse_block(&mut self) -> ExprResult {
        let id = self.consume_expect(Token::LeftBrace);
        let mut exprs = Vec::new();

        loop {
            if self.curr_token == Token::RightBrace {
                self.spans[id].end = self.skip().start;
                return Correct(Expr::Block(id, exprs));
            }

            let (expr, is_panic) = self.parse_expr(None).destruct();
            exprs.push(expr);

            if is_panic && self.curr_token != Token::RightBrace {
                self.spans[id].end = self.spans[id].start;
                return PanicMode(Expr::Block(id, exprs));
            }
        }
    }

    fn parse_infix(&mut self, mut tree: Expr, min_binding_power: u8) -> ExprResult {
        // Associativity
        const LEFT: u8 = 0;
        const RIGHT: u8 = 1;

        enum InfixType {
            Standard(BinOpKind),
            Assign,
        }
        use InfixType::{Assign, Standard};

        loop {
            let (binding_power, assoc, kind) = match self.curr_token {
                Token::Equal => (5, RIGHT, Assign),
                Token::Plus => (10, LEFT, Standard(BinOpKind::Add)),
                Token::Minus => (10, LEFT, Standard(BinOpKind::Sub)),
                Token::Star => (20, LEFT, Standard(BinOpKind::Mul)),
                Token::Slash => (20, LEFT, Standard(BinOpKind::Div)),
                Token::Percent => (20, LEFT, Standard(BinOpKind::Rem)),
                Token::DoubleStar => (30, RIGHT, Standard(BinOpKind::Pow)),
                Token::LeftParen | Token::LeftBracket | Token::Dot => {
                    todo!("Function call, indexing, dot operators not implemented.")
                }
                _ => break,
            };

            if binding_power + assoc > min_binding_power {
                let id = self.consume();
                let (expr, mut is_panic) = self.parse_expr(Some(binding_power)).destruct();

                tree = if let Standard(bin_op_kind) = kind {
                    Expr::Binary(id, bin_op_kind, Box::new(tree), Box::new(expr))
                } else if let Expr::Identifier(n) = tree {
                    // It's an assignment
                    Expr::Assign(id, n, Box::new(expr))
                } else {
                    // Assignment if tree isn't an identifier
                    is_panic = true;
                    let err_span = ast_span::get_expr_span(&tree, &self.spans);
                    let expr_span = ast_span::get_expr_span(&expr, &self.spans);
                    self.spans[id].start = err_span.start;
                    self.spans[id].end = expr_span.end;

                    self.err_consume(
                        id,
                        ParserErrorKind::ExpressionNotAssignable,
                        err_span,
                        &Self::EXPR_SYNC,
                    )
                };

                if is_panic {
                    return PanicMode(tree);
                }
            } else {
                break;
            }
        }

        Correct(tree)
    }
}
