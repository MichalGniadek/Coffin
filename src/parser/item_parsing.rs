use super::Parser;
use crate::{
    ast::{Attr, Attrs, Item, Name},
    error::ParserErrorKind,
    lexer::Token,
};

impl Parser<'_> {
    pub fn parse_item(&mut self) -> Item {
        let attrs = match self.curr_token {
            Token::HashBracket => self.parse_attributes(),
            _ => Attrs::None,
        };

        match self.curr_token {
            Token::Fun => self.parse_fun(attrs),
            _ => todo!(),
        }
    }

    fn parse_attributes(&mut self) -> Attrs {
        let brackets_id = self.consume();
        let mut attrs = vec![];

        while self.curr_token != Token::RightBracket {
            let name = match self.curr_token {
                Token::Identifier(spur) => Name {
                    id: self.consume(),
                    spur,
                },
                _ => {
                    return self.err_consume(
                        brackets_id,
                        ParserErrorKind::expected_identifier(),
                        None,
                        &Self::ITEM_SYNC,
                    )
                }
            };

            attrs.push(Attr(name, self.parse_delimited_tokens()));

            if self.curr_token == Token::Comma {
                self.skip();
            } else if self.curr_token != Token::RightBracket {
                return self.err_consume(
                    brackets_id,
                    ParserErrorKind::ExpectedToken(Token::RightBracket),
                    None,
                    &Self::ITEM_SYNC,
                );
            }
        }

        self.spans[brackets_id].end = self.skip().end - 1;
        Attrs::Ok(brackets_id, attrs)
    }

    fn parse_fun(&mut self, attrs: Attrs) -> Item {
        let fun_id = self.consume_expect(Token::Fun);

        let name = match self.curr_token {
            Token::Identifier(spur) => Name {
                id: self.consume(),
                spur,
            },
            _ => {
                return self.err_consume(
                    fun_id,
                    ParserErrorKind::expected_identifier(),
                    None,
                    &Self::ITEM_SYNC,
                )
            }
        };

        let paren_id = match self.curr_token {
            Token::LeftParen => self.consume(),
            _ => {
                return self.err_consume(
                    fun_id,
                    ParserErrorKind::ExpectedToken(Token::LeftParen),
                    None,
                    &Self::ITEM_SYNC,
                )
            }
        };

        let mut params = vec![];

        while self.curr_token != Token::RightParen {
            let field = match self.parse_field(fun_id) {
                Ok(f) => f,
                Err(err) => return err,
            };

            params.push(field);

            if self.curr_token == Token::Comma {
                self.skip();
            } else if self.curr_token != Token::RightParen {
                return self.err_consume(
                    fun_id,
                    ParserErrorKind::ExpectedToken(Token::RightParen),
                    None,
                    &Self::ITEM_SYNC,
                );
            }
        }

        self.spans[paren_id].end = match self.curr_token {
            Token::RightParen => self.skip().end - 1,
            _ => {
                return self.err_consume(
                    fun_id,
                    ParserErrorKind::ExpectedToken(Token::RightParen),
                    None,
                    &Self::ITEM_SYNC,
                )
            }
        };

        let ret = if self.curr_token == Token::Arrow {
            let arrow_id = self.consume();

            let ttpe = match self.curr_token {
                Token::Identifier(spur) => Name {
                    id: self.consume(),
                    spur,
                },
                _ => {
                    return self.err_consume(
                        fun_id,
                        ParserErrorKind::expected_identifier(),
                        None,
                        &Self::ITEM_SYNC,
                    )
                }
            };

            Some((arrow_id, ttpe))
        } else {
            None
        };

        let body = match self.curr_token {
            Token::LeftBrace => self.parse_expr(None).expr(),
            _ => self.err_consume(
                fun_id,
                ParserErrorKind::ExpectedToken(Token::LeftBrace),
                None,
                &Self::ITEM_SYNC,
            ),
        };

        Item::Fun {
            fun_id,
            attrs,
            name,
            paren_id,
            params,
            ret,
            body,
        }
    }
}
