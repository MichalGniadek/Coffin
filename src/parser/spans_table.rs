use std::ops::{Index, IndexMut};
use logos::Span;
use crate::ast::Id;

#[derive(Debug, Clone)]
pub struct SpansTable(Vec<Span>);

impl SpansTable {
    pub fn new() -> Self {
        Self(vec![])
    }

    pub fn push(&mut self, value: Span) -> Id {
        self.0.push(value);
        Id(self.0.len() as usize - 1)
    }
}

impl Index<Id> for SpansTable {
    type Output = Span;

    fn index(&self, index: Id) -> &Self::Output {
        &self.0[index.0 as usize]
    }
}

impl IndexMut<Id> for SpansTable {
    fn index_mut(&mut self, index: Id) -> &mut Self::Output {
        &mut self.0[index.0 as usize]
    }
}