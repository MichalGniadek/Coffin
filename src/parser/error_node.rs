use crate::{
    ast::{self, Attrs, Id},
    error::ParserError,
};
use super::expr_parsing::ExprResult;

pub struct ErrorNode(pub Id, pub ParserError);

impl From<ErrorNode> for ast::Item {
    fn from(err: ErrorNode) -> Self {
        Self::Error(err.0, err.1)
    }
}

impl From<ErrorNode> for ast::Expr {
    fn from(err: ErrorNode) -> Self {
        Self::Error(err.0, err.1)
    }
}

impl From<ErrorNode> for ExprResult {
    fn from(err: ErrorNode) -> Self {
        Self::PanicMode(err.into())
    }
}

impl From<ErrorNode> for ast::Attrs {
    fn from(err: ErrorNode) -> Self {
        Attrs::Error(err.0, err.1)
    }
}
