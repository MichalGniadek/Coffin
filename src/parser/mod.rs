mod error_node;
mod expr_parsing;
mod item_parsing;

use crate::{
    ast::{Expr, Id, Spans, UntypedAst},
    error::{ParserError, ParserErrorKind},
    lexer::Token,
};
use logos::{Lexer, Span, SpannedIter};
use std::iter::Peekable;

pub fn parse(lexer: Lexer<'_, Token>) -> UntypedAst {
    let mut parser = Parser {
        lexer: lexer.spanned().peekable(),
        spans: Spans::new(),
    };

    let mut items = vec![];

    while parser.peek() != Token::EOF {
        items.push(parser.parse_item());
    }

    UntypedAst {
        items,
        spans: parser.spans,
    }
}

struct Parser<'a> {
    lexer: Peekable<SpannedIter<'a, Token>>,
    spans: Spans,
}

enum ParserResult {
    Ok(Expr),
    Panic(Expr),
}
use ParserResult::{Ok, Panic};

use self::error_node::ErrorNode;

impl ParserResult {
    fn expr(self) -> Expr {
        match self {
            Ok(e) => e,
            Panic(e) => e,
        }
    }
    fn destruct(self) -> (Expr, bool) {
        match self {
            Ok(e) => (e, false),
            Panic(e) => (e, true),
        }
    }
}

const fn concat<const A: usize, const B: usize, const C: usize>(
    a: [Token; A],
    b: [Token; B],
) -> [Token; C] {
    // Check lengths
    let _ = [0; 1][A + B - C];

    let mut out = [Token::EOF; C];

    let mut i = 0;

    while i < A {
        out[i] = a[i];
        i += 1;
    }

    while i < A + B {
        out[i] = b[i - A];
        i += 1;
    }

    out
}

impl Parser<'_> {
    const ITEM_SYNC: [Token; 2] = [Token::Fun, Token::EOF];
    const EXPR_SYNC: [Token; 3] = concat(Self::ITEM_SYNC, [Token::RightBrace]);

    /// Peeks the next token.
    fn peek(&mut self) -> Token {
        self.lexer.peek().map_or(Token::EOF, |(t, _)| t.clone())
    }

    /// Peeks the next span.
    fn peek_span(&mut self) -> Span {
        self.lexer.peek().map_or(0..0, |(_, s)| s.clone())
    }

    /// Advances to the next token and returns the previous span.
    fn skip(&mut self) -> Span {
        let span = self.peek_span();
        self.lexer.next();
        span
    }

    /// Advances to the next token, add span to the Spans and return the
    /// corresponding id.
    fn consume(&mut self) -> Id {
        let span = self.peek_span();
        self.lexer.next();
        self.spans.push(span)
    }

    /// Advances to the next token, add span to the Spans and return the
    /// corresponding id. Panics if the next token isn't the correct.
    fn consume_expect(&mut self, token: Token) -> Id {
        assert!(
            self.peek() == token,
            "Compiler error: expected '{:?}' token.",
            token
        );
        self.consume()
    }

    /// Consumes tokens until it get to sync token and return error variant of an AST type.
    fn err_consume<T: From<ErrorNode>>(&mut self, kind: ParserErrorKind, tokens: &[Token]) -> T {
        let span = self.peek_span();
        let id = self.spans.push(span);
        self.err_consume_append(id, kind, tokens)
    }

    /// Consumes tokens until it get to sync token and return error variant of an AST type.
    /// The id of the node will be the same as the passed one.
    fn err_consume_append<T: From<ErrorNode>>(
        &mut self,
        id: Id,
        kind: ParserErrorKind,
        tokens: &[Token],
    ) -> T {
        let err_span = self.peek_span();

        while let Some((token, span)) = self.lexer.peek() {
            if !tokens.contains(token) {
                self.spans[id].end = span.end;
                self.lexer.next();
            } else {
                break;
            }
        }

        ErrorNode(id, ParserError(kind, err_span)).into()
    }
}
