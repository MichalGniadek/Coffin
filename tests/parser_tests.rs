use coffin2::parser;
use coffin2::{ast::Visitor, lexer::Token, pretty_print::PrettyPrint};
use insta::{assert_snapshot, glob};
use logos::Logos;
use std::fs;

#[test]
fn insta() {
    glob!(r"shaders\*.coff", |path| {
        let code = fs::read_to_string(path).unwrap();
        let lexer = Token::lexer(&code);
        let ast = parser::parse(lexer);
        let mut print = PrettyPrint::new(&ast.spans);
        assert_snapshot!(ast
            .items
            .iter()
            .map(|i| print.visit_item(i))
            .reduce(|a, b| format!("{}\n{}", a, b))
            .unwrap());
    });
}
