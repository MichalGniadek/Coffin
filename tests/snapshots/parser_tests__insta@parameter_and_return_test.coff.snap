---
source: tests/parser_tests.rs
expression: "DebugPrint::visit(&ast, Some(&rodeo), None, None)"

---
fun asd (parm1 : Tt1, param2 : Tt2) -> i32 {
	
}
